### PR TITLE
T343 schoolie

### PR DESCRIPTION
**Pull request by kerchner**
This PR includes a rake task to migrate existing data, so it is recommended to test not only on:
1. A clean instance, but also:
2. An instance with existing ETDs, as well as at least one non-ETD Work.  These ETDs should (as all current ETDs do) have `resource_type` set to `Thesis or Dissertation` and should have `degree` populated.  Populate at least one ETD with a masters flavored degree, and at least one ETD with a doctoral flavored degree.  Ensure that at least one of the ETDs has 2 or more keywords populated.

### On an instance with existing ETDs

See prerequisite data above.

1. Run the `gwss:enumerate_degree_types` rake task and observe that the task prints out unique `degree` values, with a count of each
2. Run the `gwss:reassign_etd_resource_types` task and observe that:
- the rake task's messages contained a "reassigned" message for each ETD, and no "degree name not found" messages.
- view in the UI that every ETD's `resource_type` was reassigned to either `Master's Thesis` or `Dissertation`

### On a new instance

1.  Create an ETD that is has `resource_type` set to `Master's Thesis`, and an ETD with `resource_type` set to `Dissertation`.  Ensure that at one of them has 2 or more keywords. 
2. Create at least one non-ETD work.

### On both new and migrated instances

1. View the Page Source for an ETD.  Observe that the `<head>` tag contains a block similar to this:

```
<!-- Schoolie for Google Scholar metadata (replaces :gscholar_meta) -->
<meta name="citation_institution" value="George Washington University">
<meta name="citation_title" value="Test ETD">
<meta name="citation_author" value="Kerchner, Daniel">
<meta name="citation_type" value="Master&#39;s Thesis">
<meta name="dc.type" value="Master&#39;s Thesis">
<meta name="citation_keywords" value="keyword1">
<meta name="citation_keywords" value="keyword2">
<meta name="citation_pdf_url" value="http://ec2-35-91-19-226.us-west-2.compute.amazonaws.com/downloads/6969z076v">
```

2. Scrape the URL from the `citation_pdf_url` meta tag above and confirm that the browser downloads (and does not display in-browser) the expected PDF.

3. View the Page Source for a non-ETD work.  Observe that  `<head>` tag contains a block similar to this:

```
<!-- Schoolie for Google Scholar metadata (replaces :gscholar_meta) -->
```
with none of the other tags shown in the previous step.

### Checking the sitemap

1.  Confirm that the `schoolie:sitemap` rake task creates the sitemap and that it looks similar to this:

```
<?xml version="1.0"?>
<urlset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
  <url>
    <loc>https://scholarspace.library.gwu.edu/etd/9306sz28s</loc>
    <lastmod>2024-02-29T16:59:42Z</lastmod>
  </url>
  <url>
    <loc>https://scholarspace.library.gwu.edu/etd/3069xz28s</loc>
    <lastmod>2024-02-29T15:56:00Z</lastmod>
  </url>
</urlset>
```

Note that the URL is going to be `scholarspace.library.gwu.edu` not your local server.

2. Adjusting for the URL, confirm that the URLs lead to the landing pages of each ETD (not the download URLs)
3. Confirm that the `lastmod` time seems to correspond to the time the work was last modified
4. Confirm that non-ETDs are _not_ included in the sitemap.   I know that's weird, but right now we're trying to avoid ticking off Google Scholar because we have "mixed" content - scholarly and non-scholarly.  Hopefully this is temporary.
5. Confirm that the sitemap is regenerated by the scheduled sitemap regeneration cron job.  Perhaps delete it first to confirm, or you can rely on the timestamp.  Either wait until tomorrow, or adjust the schedule as per https://github.com/gwu-libraries/scholarspace-hyrax/pull/487

### Known issues

I cannot seem to get the sitemap generation to trigger using the `gwss:sitemap_queue_generate` rake task.  The only difference is that I've replaced the contents of `SitemapRegenerateJob` to call the new rake task instead.  Should I be concerned?

### Thoughts on the implementation

I am not crazy about having added `scholarly?` into the presenters rather than the work models, but I couldn't find a clean way to be able to ask the work's `scholarly?` from the HTML.   The reason that I think ultimately it belongs on the work, is so that we might add more logic to the function that can answer true/false perhaps depending on the values of certain properties.   Since we don't yet have such a requirement, this is something we can refactor later.

Any feedback on the code/implementation is most welcome.

### CODE REVIEW

File: app/gw_work.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Generated via
```
Comment (alepbloyd) [inline app/gw_work.rb:1]: @kerchner Should this file be here? There's a copy in `/app/models/gw_work.rb` which makes sense, but not sure why we would want this in the `/app` root folder.

File: app/gw_work.rb
Comment (alepbloyd) [inline app/gw_work.rb:1]: @kerchner Should this file be here? There's a copy in /app/models/gw_work.rb which makes sense, but not sure why we would want this in the /app root folder.

Comment (kerchner) [reply app/gw_work.rb:1]: Indeed it should not.  I've removed it.

**Comment by alepbloyd** [comment]
I ran the rake tasks on the backup data I have (which is from like september-ish 2023). I was able to run the gwss:enumerate_degree_types task and got a list of results, but there are these two categories that look like they have empty strings or null values.
image

Digging back through the logs, and I did find this `Degree name   not found` line.
image

Probably related to that, when I ran the `gwss:reassign_etd_resource_types` task - it ran fine for probably about 15 minutes or so, and then I got a `NoMethodError: undefined method 'upcase' for nil:NilClass` for line 341 in the task.
image

Might need to add in a way to handle these few works with null values in the field, since it looks like we have those in prod as well: 
image

**Comment by kerchner** [comment]
@alepbloyd Got it, I'll fix to skip those (leave as `Thesis or Dissertation`) but identify them.   This will allow us to target those three ETDs in production to manually correct their `degree` values as needed, and set `resource_type` accordingly.  

Will keep you posted when I update the rake task.

**Comment by kerchner** [comment]
@alepbloyd Please re-test `gwss:reassign_etd_resource_types`

**Comment by alepbloyd** [comment]
@kerchner

update on testing this:

I re-restored my scholarspace instance for the backups I have, and then ran the `gwss:enumerate_degree_types` task - worked fine. After that, I started the `gwss:reassign_etd_resource_types` task. It took about 4.5 hours to run on the backup data, but didn't get any errors or anything. Looking through the UI, all looks good. 

I did run into a situation where two or three minutes after the task finished, I seemed to get kicked out of my EC2 server and couldn't reconnect until I went into the AWS console and rebooted it. Now I can connect again fine, so not really sure what that was about - since it ran fine for hours and didn't have any issue until after the fact.

I still need to look at the sitemap aspects of this, aiming to finish today/tomorrow.

**Comment by alepbloyd** [comment]
@kerchner 

- Confirmed that I am seeing the <meta> tags for ETDs in the page source, looks like it worked with multiple keywords as well, and not seeing them for non-ETDs. 
- For the `citation_pdf_url`, I did need to change it from `http` to `https`, though I would imagine that's just something with my SSL setup on my dev server.
- Regenerated the sitemap using `bundle exec rails schoolie:sitemap RAILS_ENV=production` on my server, confirmed that looks correct and the landing page links are working (with URL replaced). 
  - We can actually see the ETDs that didn't have their `Resource Type` updated, because they are the only ones in the sitemap where the <lastmod> doesn't correspond with the time I ran the `reassign_resource_types` task.
- Confirmed that non-ETDs are not appearing in the sitemap. 
- I didn't have any issues when I ran the `gwss:sitemap_queue_generate` task. Seemed to work fine for me, and logs point to it running on Sidekiq, so that seems correct. 

I just deleted the sitemap, and will check back on it tomorrow morning to verify if it regenerated as per the cron job. 

So beyond that one pending aspect, everything seems to be working!

**Comment by kerchner** [comment]
Thanks @alepbloyd !  Keep me posted on the sitemap.

**Comment by kerchner** [comment]
I'm going to remove `sitemap` from `Gemfile` and also remove `config/sitemap.rb`.  We probably will want to add back more pages to the sitemap in the near future, but not until we've succeeded in having Google Scholar index ETDs.

**Review by alepbloyd — APPROVED** [review]
Looks good!

**Pull request merged**